### PR TITLE
created security features for api

### PR DESCRIPTION
**Pull request by johnayinde**
added  secret key to Api  a UUID string to protect the api from being accessed by an unauthorized user
"Api-Secret-Key" is added on every request made to the api with the secret key as value

### CODE REVIEW

File: src/subscription/subscription.service.ts [hunk 1]
```diff
@@ -219,7 +223,17 @@ export class SubscriptionService {
     }
     // we need too check that the name, data type and requirements foe each property in the
     //endpoints.payload are met explicitly
-    
+
+    // verify if the api secret key value is the same as the api secret
+    if (api.secretKey != uniqueApiSecurityKey) {
```
Comment (PhilemonBrain) [inline src/subscription/subscription.service.ts:228]: this unique api key is not for us. It's meant to be attached to the headers only

File: src/entities/api.entity.ts [hunk 2]
```diff
@@ -56,6 +56,9 @@ export class Api extends SharedEntity {
   @Column()
   profileId: string;
 
+  @Column()
```
Comment (PhilemonBrain) [inline src/entities/api.entity.ts:59]: there is no migration file for these changes on the entity schema.

File: src/entities/api.entity.ts [hunk 3]
```diff
@@ -56,6 +56,9 @@ export class Api extends SharedEntity {
...
+  secretKey: string;
+
```
Comment (PhilemonBrain) [inline src/entities/api.entity.ts:61]: What will happen to already existing entities when u add this changes

File: src/subscription/subscription.controller.ts [hunk 4]
```diff
@@ -34,22 +35,35 @@ export class SubscriptionController {
...
     @Body() subscriptionApiCall: SubscriptionApiCallDto,
-  ):Promise<Ok<Object>>{
+    @Req() request: Request,
+  ): Promise<Ok<Object>> {
     // const verifySubscription = await this.subscriptionService.verifySub(verifysub)
-    const verifySubscription = await this.subscriptionService.makeSubscriptionRequest(token, subscriptionApiCall)
-    return ZapiResponse.Ok(verifySubscription, 'user is subcribed to this api', '200')
+
+    // Get the secret key from the request header "Api-Secret-Key"
+    const uniqueApiSecurityKey = request.headers['api-secret-key'];
```
Comment (PhilemonBrain) [inline src/subscription/subscription.controller.ts:47]: there wont be such a header in this request and change the header name to X-ZAPI-Proxy-Secret

**Review by PhilemonBrain — CHANGES_REQUESTED** [review]

**Comment by johnayinde** [comment]
New changes has been made 
Migration file generated and request header is sent on ever request made to an api

File: src/database/1658490667193-migration.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import { MigrationInterface, QueryRunner } from "typeorm";
+
+export class migration1658490667193 implements MigrationInterface {
```
Comment (PhilemonBrain) [inline src/database/1658490667193-migration.ts:3]: @johnayinde this migration is wrong

Comment (PhilemonBrain) [reply src/database/1658490667193-migration.ts:3]: do u understand migrations?

**Review by PhilemonBrain — CHANGES_REQUESTED** [review]

File: src/database/1658490667193-migration.ts (hunk 5 above)
Comment (johnayinde) [reply src/database/1658490667193-migration.ts:3]: @PhilemonBrain 
I don’t really understand it that well

 I went through the Typeorm doc and saw the same script command on the package.json file to generate the migration file **add-migration**

Comment (PhilemonBrain) [reply src/database/1658490667193-migration.ts:3]: @johnayinde  see if u can schedule a meet with @SunmadeBabafemi or @andyjrII to put u through

**Review by PhilemonBrain — APPROVED** [review]

**Pull request merged**